### PR TITLE
Build consume stock

### PR DESCRIPTION
**Pull request by SchrodingersGat**
Closes https://github.com/inventree/InvenTree/issues/3012

Untracked stock items are now "consumed" by a build order, rather than being "deleted" from the database.

### TODO

- [x] Unit testing
- [x] Documentation

### CODE REVIEW

File: docs/docs/terminology.md [hunk 1]
```diff
@@ -6,23 +6,27 @@ title: Terminology
 
 There are different systems in the industry for the management of getting, storing and making parts. An overview what they are for and what the acronyms mean.
 
+**InvenTree** is mainly focused on [**IMS**](#inventory-management-system-ims) and [**PLM**](#part-library-management-plm) functionality.
```
Comment (matmair) [inline docs/docs/terminology.md:9]: Comment: The scope is growing ;-)

File: InvenTree/InvenTree/tasks.py [hunk 2]
```diff
@@ -563,6 +563,16 @@ def send_email(subject, body, recipients, from_email=None, html_message=None):
...
 
+    import InvenTree.ready
+
+    if InvenTree.ready.isImportingData():
+        # If we are importing data, don't send emails
+        return
+
+    if InvenTree.ready.isInTestMode():
+        # If we are running in test mode, don't send emails
+        return
```
Comment (matmair) [inline InvenTree/InvenTree/tasks.py:574]: Question: Should we perhaps test if email delivery is working somewhere?

File: InvenTree/build/templates/build/detail.html [hunk 3]
```diff
@@ -337,11 +360,9 @@ <h4>{% trans "Build Notes" %}</h4>
             build: {{ build.id }},
             is_building: false,
         },
-        groupByField: 'location',
```
Comment (matmair) [inline InvenTree/build/templates/build/detail.html:340]: Question: Will this change the default behavior and is that intended?
Suggestion: If yes: Should this PR be marked as breaking?

File: InvenTree/stock/models.py [hunk 4]
```diff
@@ -755,6 +756,14 @@ def get_part_name(self):
         related_name='build_outputs',
     )
 
+    consumed_by = models.ForeignKey(
+        'build.Build', on_delete=models.CASCADE,
```
Comment (matmair) [inline InvenTree/stock/models.py:760]: Question: Is it intended to delete the stock items completely when the build is deleted?

**Review by matmair — COMMENTED** [review]
Nice addition; Does it make sense to update the selection dropdowns with a separate field for consumed items?

File: InvenTree/InvenTree/tasks.py (hunk 2 above)
Comment (SchrodingersGat) [reply InvenTree/InvenTree/tasks.py:574]: I'm going to revert this particular line anyway, it's breaking CI :)

File: InvenTree/build/templates/build/detail.html (hunk 3 above)
Comment (SchrodingersGat) [reply InvenTree/build/templates/build/detail.html:340]: No, stock-item grouping has long been removed! This is just a cleanup of legacy code

File: InvenTree/stock/models.py (hunk 4 above)
Comment (SchrodingersGat) [reply InvenTree/stock/models.py:760]: Yes, that's correct.

The logic is:

1. Currently, untracked items are removed from the database (deleted) once consumed by a build
2. New behaviour is that the untracked items are marked as "consumed_by" - but kept in the database
3. The "consumed_by" field is what tells the database that these items are unavailable for consumption
4. When the build is deleted, I think the "cleanest" thing to do is to remove those items (as opposed to returning them to stock)

Open to alternative suggestions though :)

**Comment by SchrodingersGat** [comment]
> Nice addition; Does it make sense to update the selection dropdowns with a separate field for consumed items?

Do you mean the filters in the stock tables? Or something else?

**Comment by SchrodingersGat** [comment]
I don't think that the user should manually mark the stock items as "consumed by", or have the ability to do this via the front-end modals.

The stock items can always be manually edited in the admin interface

**Pull request merged**